src/chart/generateCategoricalChart.tsx

```diff
@@ -288,87 +313,105 @@ const getAxisMapByAxes = (
 
     let domain, duplicateDomain, categoricalDomain;
 
-    if (dataKey) {
-      // has dataKey in <Axis />
-      domain = getDomainOfDataByKey(displayedData, dataKey, type);
-
-      if (type === 'category' && isCategorical) {
-        // the field type is category data and this axis is catrgorical axis
-        const duplicate = hasDuplicate(domain);
-
-        if (allowDuplicatedCategory && duplicate) {
-          duplicateDomain = domain;
-          // When category axis has duplicated text, serial numbers are used to generate scale
-          domain = _.range(0, len);
-        } else if (!allowDuplicatedCategory) {
-          // remove duplicated category
-          domain = parseDomainOfCategoryAxis(child.props.domain, domain, child).reduce(
-            (finalDomain: any, entry: any) => (finalDomain.indexOf(entry) >= 0 ? finalDomain : [...finalDomain, entry]),
-            [],
+    /*
+     * This is a hack to short-circuit the domain creation here to enhance performance.
+     * Usually, the data is used to determine the domain, but when the user specifies
+     * a domain upfront (via props), there is no need to calculate the domain start and end,
+     * which is very expensive for a larger amount of data.
+     * The only thing that would prohibit short-circuiting is when the user doesn't allow data overflow,
+     * because the axis is supposed to ignore the specified domain that way.
+     */
+    if (isDomainSpecifiedByUser(child.props.domain, allowDataOverflow, type)) {
+      domain = parseSpecifiedDomain(child.props.domain, null, allowDataOverflow);
+    }
+
+    // we didn't create the domain from user's props above, so we need to calculate it
+    if (!domain || domain.length === 0) {
```

Comment:
this breaks highlight and zoom functionality - confirmed. Commented out and tested, functionality works again.

Reply:
Looks like if we have a categorical chart we still need to find and set a categorical domain, else the `categoricalDomain` is undefined and this happens